### ISSUE TITLE
oneDNN Pytorch inference time issue

### ISSUE THREAD

**Issue by rubende** [oneapi-src/oneDNN#1645]
Hello,

I am trying to optimize a CPU model using oneDNN, following [this tutorial](https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html#use-onednn-graph-with-torchscript-for-inference)

The situation is that I observe a huge difference in performance by inference between the model generated for CPU with PyTorch and the model generated directly with the oneDNN library version 1.6 (300 microseconds in the first case and less than 20 microseconds in the second).

The procedure I am following with the model in Python to prepare it for CPU is this:

```
model = model.to(“cpu”)
torch.jit.enable_onednn_fusion(True)
model.eval()
model = torch.jit.trace(model, torch.rand(shape))
model = torch.jit.freeze(model _scripted)
torch.jit.save(model, path)
```

After that, in C++, I use it as follows:

```
module = torch::jit::load(path);
output = module.forward(tensor_inputs).toTensor();
```

Not getting the expected results, I have also tried with the library intel-extension-for-pytorch from Intel, including in the previous code this line before “torch.jit.trace”:

`model = ipex.optimize(model, dtype=torch.float32)`

The inference time is, also, around 300 microseconds. Also, performing the command " ldd" on the compiled executable file, and comparing with [this tutorial](https://github.com/intel/intel-extension-for-pytorch/blob/master/docs/tutorials/examples.md), I notice that I am missing the file “libdnnl_graph.so”, which I am not clear where it should come from because I cannot find it in any of the aforementioned libraries.

Any idea or suggestion of what could be happening? Does the optimization process for CPU seem correct to you? Is this huge difference in performance normal?

Thank you.

**Comment by vpirogov** [comment]
Considering that this is a model level performance issue I'd suggest you report it to [IPEX](https://github.com/intel/intel-extension-for-pytorch) or [Pytorch](https://github.com/pytorch/pytorch). 

There are many factors that may lead to the differences in performance. You may compare how oneDNN is used in Pytorch/IPEX and your implementation using [oneDNN verbose mode](https://oneapi-src.github.io/oneDNN/v3.1/dev_guide_verbose.html). Running the code with `DNNL_VERBOSE=1` will produce information about all oneDNN calls.

The `libdnnl_graph.so` is a part of [oneDNN Graph technical preview](https://github.com/oneapi-src/oneDNN/releases/tag/graph-v0.9). This functionality is included into `libdnnl.so` starting from [v3.1](https://github.com/oneapi-src/oneDNN/releases/tag/v3.1).

+@yehudaorel, @louie-tsai

**Comment by rubende** [comment]
Hello,

thank you very much for your reply. 

With `DNNL_VERBOSE=1`  I have checked that I am using oneDNN, version 3.0.1.
In this version, I understand that I should see the file `libdnnl_graph.so`. 

Should I compile it and try to link it? I am trying to compile it with cmake, but I am getting some error which is not specified. Is there any document explaining how to perform these steps?

`
-- CMAKE_BUILD_TYPE is unset, defaulting to Release
-- The C compiler identification is GNU 9.4.0
-- The CXX compiler identification is GNU 9.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at cmake/build_oneDNN.cmake:45 (add_subdirectory):
  The source directory

    /home/rdelgado/workspace/oneDNN-graph-v0.9/third_party/oneDNN

  does not contain a CMakeLists.txt file.
Call Stack (most recent call first):
  cmake/build_oneDNN.cmake:48 (build_onednn)
  CMakeLists.txt:142 (include)


-- Looking for /home/rdelgado/workspace/oneDNN-graph-v0.9/include/oneapi/dnnl/dnnl_graph.h
-- Looking for /home/rdelgado/workspace/oneDNN-graph-v0.9/include/oneapi/dnnl/dnnl_graph.h - found
-- Looking for /home/rdelgado/workspace/oneDNN-graph-v0.9/include/oneapi/dnnl/dnnl_graph_types.h
-- Looking for /home/rdelgado/workspace/oneDNN-graph-v0.9/include/oneapi/dnnl/dnnl_graph_types.h - found
-- Looking for C++ include /home/rdelgado/workspace/oneDNN-graph-v0.9/include/oneapi/dnnl/dnnl_graph.hpp
-- Looking for C++ include /home/rdelgado/workspace/oneDNN-graph-v0.9/include/oneapi/dnnl/dnnl_graph.hpp - found
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE)
-- Cannot find Doxygen package
-- Found Git: /usr/bin/git (found version "2.25.1")
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
-- Found OpenMP_C: -fopenmp (found version "4.5")
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- DNNL_GRAPH_BUILD_FOR_CI is set to be OFF
-- Compiling oneDNN Graph with CPU runtime OMP support
-- Compiling oneDNN Graph with GPU runtime NONE support
-- Graph compiler backend is disabled.
-- Set version definitions to /home/rdelgado/workspace/oneDNN-graph-v0.9/src/utils/verbose.cpp
-- Compiled partition cache is enabled
-- Configuring incomplete, errors occurred!
See also "/home/rdelgado/workspace/oneDNN-graph-v0.9/build/CMakeFiles/CMakeOutput.log".
`

CMakeOutput.log does not include information about the error produced.

Thank you.

**Comment by TaoLv** [comment]
@rubende  do you see files in the folder `/home/rdelgado/workspace/oneDNN-graph-v0.9/third_party/oneDNN` ?  It seems you need to run `git submodule update --recursive` somewhere to pull the submodule files.

**Comment by rubende** [comment]
@TaoLv 

No, I don't see any file in `/home/rdelgado/workspace/oneDNN-graph-v0.9/third_party/oneDNN`

Command `git submodule update --recursive` not solve it.

About IPEX, I don't see any reference to "libdnnl_graph.so" in its compilation process. Any suggestions on how I could check if my model is using that library? In parallel, I am going to ask the IPEX repository about this issue.

Thank you.

**Comment by TaoLv** [comment]
> Command git submodule update --recursive solve it.

@rubende  Try `git submodule update --recursive --init` in the folder oneDNN-graph-v0.9. Please share the log message if it still does not pull code to the third_party/oneDNN/ folder.

For IPEX build, yes, it would be better to raise question in IPEX repository. Tagging IPEX developers for cross reference, @chunyuan-w @sanchitintel

**Comment by rubende** [comment]
@TaoLv Compilation problem solved with `git submodule update --recursive --init`. Thank you.

I opened a new issue in IPEX repository: https://github.com/intel/intel-extension-for-pytorch/issues/348
